### PR TITLE
Robustify RefCountingDisposable

### PR DESCRIPTION
**Pull request by benaadams**
Currently if 2 threads simultaniously try to Aquire a disposed object one may suceed

1. Starting value = -1
2. Thread 1: increment = 0
3. Thread 2: increment = 1
4. Thread 1, fail: decrement = 0
5. Thread 2, suceeded => invalid state

### CODE REVIEW

File: src/Paprika/Utils/RefCountingDisposable.cs [hunk 1]
```diff
@@ -11,13 +12,12 @@ public abstract class RefCountingDisposable : IDisposable
     private const int Initial = 1;
     private const int NoAccessors = 0;
     private const int Disposing = -1;
-    private const int DisposingBarrier = 0;
 
-    private int _leases;
+    private PaddedValue _leases;
```
Comment (Scooletz) [inline src/Paprika/Utils/RefCountingDisposable.cs:16]: I see the padding here. I know what it does. The thing is that this is the abstract class and what I'm thinking is whether is needed. WDYT?

File: src/Paprika/Utils/RefCountingDisposable.cs [hunk 2]
```diff
@@ -57,23 +73,65 @@ protected bool TryAcquireLease()
...
-            if (Interlocked.CompareExchange(ref _leases, Disposing, NoAccessors) == NoAccessors)
+            var prev = Interlocked.CompareExchange(ref _leases.Value, current - 1, current);
+            if (prev != current)
+            {
+                current = prev;
+                // Add PAUSE instruction to reduce shared core contention
+                Thread.SpinWait(1);
+                continue;
+            }
+            if (prev == 1)
```
Comment (Scooletz) [inline src/Paprika/Utils/RefCountingDisposable.cs:94]: Magic const. This could be 
```suggestion
            if (prev == Initial)
```

**Review by Scooletz — COMMENTED** [review]
Good to go after clarifications @benaadams

File: src/Paprika/Utils/RefCountingDisposable.cs (hunk 2 above)
Comment (benaadams) [reply src/Paprika/Utils/RefCountingDisposable.cs]: Changed

File: src/Paprika/Utils/RefCountingDisposable.cs (hunk 1 above)
Comment (benaadams) [reply src/Paprika/Utils/RefCountingDisposable.cs:16]: If it is hevily used changing its value causes anything on its cache line to be invalidated for other CPUs and forces them to be reloaded.

So for example the pointer of `HashSet<ulong> _bloom` would keep getting invaliated in `BlockState`; or pointer of `PooledSpanDictionary _dict` in `ChildCommit`

Comment (benaadams) [reply src/Paprika/Utils/RefCountingDisposable.cs:16]: Depends if worried about the memory usage by the extra padding or not

**Review by Scooletz — APPROVED** [review]

**Pull request merged**